src/gui/plugins/visualize_forces/VisualizeForces.cc

```diff
@@ -0,0 +1,563 @@
+
+#include <gz/gui/Application.hh>
+#include <gz/gui/Conversions.hh>
+#include <gz/gui/GuiEvents.hh>
+#include <gz/gui/MainWindow.hh>
+
+#include <gz/math/Pose3.hh>
+#include <gz/math/Vector3.hh>
+
+#include <gz/msgs/entity_wrench_map.pb.h>
+
+#include <gz/plugin/Register.hh>
+
+#include <gz/rendering/ArrowVisual.hh>
+#include <gz/rendering/RenderEngine.hh>
+#include <gz/rendering/RenderingIface.hh>
+#include <gz/rendering/RenderTypes.hh>
+#include <gz/rendering/Scene.hh>
+
+#include <gz/transport/Node.hh>
+
+#include "gz/sim/components/EntityWrench.hh"
+#include "gz/sim/components/Pose.hh"
+#include "gz/sim/Link.hh"
+#include "gz/sim/World.hh"
+
+namespace gz
+{
+namespace sim
+{
+inline namespace GZ_SIM_VERSION_NAMESPACE
+{
+/// \brief Private data class for VisualizeForces
+class VisualizeForcesPrivate
+{
+  /// \brief Model used to synchronize with the GUI
+  public: ForceListModel model;
+
+  /// \brief Node for receiving incoming requests.
+  public: transport::Node node;
+
+  /// \brief Queue for incoming messages
+  public: std::queue<std::tuple<math::Pose3d, msgs::EntityWrench>> queue;
+
+  /// \brief Mutex for queue and other variables shared with render thread.
+  public: std::mutex mutex;
+
+  /// \brief Set of markers already drawn.
+  public: std::unordered_set<std::string> onScreenMarkers;
+
+  /// \brief Pointer to the scene.
+  public: rendering::ScenePtr scene;
+
+  /// \brief Visuals for each force marker.
+  public: std::unordered_map<
+      std::string, rendering::ArrowVisualPtr> visuals;
+
+  /// \brief Set the scale. A scale of 1 => force of 1N has a marker length 1m.
+  public: double scale{1.0};
+
+  /// \brief System update period calculated from <update_rate>
+  public: std::chrono::steady_clock::duration updatePeriod{0};
+
+  /// \brief Last system update simulation time
+  public: std::chrono::steady_clock::duration lastUpdateTime{0};
+
+
+  /// \brief Destructor
+  public: ~VisualizeForcesPrivate();
+
+  /// \brief Constructor
+  public: VisualizeForcesPrivate();
+
+  /// \brief Get the wrench visuals - runs on GUI thread.
+  /// \param[in] _ecm - the ecm.
+  public: void RetrieveWrenchesFromEcm(EntityComponentManager &_ecm);
+
+  /// \brief Render force visuals - runs on render thread.
+  public: void OnRender();
+
+  public: void AddVisual(const std::string &_ns,
+      const math::Color &_color);
+
+  public: void RemoveVisual(const std::string &_ns);
+
+  public: void ClearVisuals();
+};
+
+/////////////////////////////////////////////////
+VisualizeForcesPrivate::~VisualizeForcesPrivate()
+{
+  this->ClearVisuals();
+}
+
+/////////////////////////////////////////////////
+VisualizeForcesPrivate::VisualizeForcesPrivate() = default;
+
+/////////////////////////////////////////////////
+void VisualizeForcesPrivate::RetrieveWrenchesFromEcm(
+    EntityComponentManager &_ecm)
+{
+  std::lock_guard<std::mutex> lock(mutex);
+  {
+    _ecm.Each<components::WorldPose, components::EntityWrenchMap>(
```

Comment:
For gravity we could add a search for inertial components and then multiply by the gravity vector defined in the world entity.

Reply:
Sounds good - but for a future PR right?

Reply:
Yep.